src/coreclr/jit/lclvars.cpp

```diff
@@ -4527,13 +4529,41 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
     const bool oldLvaGenericsContextInUse = lvaGenericsContextInUse;
     lvaGenericsContextInUse               = false;
 
+#if FEATURE_LOOP_ALIGN
+    alignBlocksList* alignBB = nullptr;
+#endif
+
     JITDUMP("\n*** lvaComputeRefCounts -- explicit counts ***\n");
 
     // Second, account for all explicit local variable references
     for (BasicBlock* const block : Blocks())
     {
         if (block->IsLIR())
         {
+#if FEATURE_LOOP_ALIGN
+            if (calculateAlign)
+            {
+                // Track the blocks that needs alignment
+                // except if it is first block, because currently adding padding
+                // in prolog is not supported
+                if (opts.compJitHideAlignBehindJmp && block->isLoopAlign() && (block != fgFirstBB))
+                {
+                    alignBlocksList* curr = new (this, CMK_FlowList) alignBlocksList(block);
+
+                    if (alignBBLists == nullptr)
+                    {
+                        alignBBLists = curr;
+                    }
+                    else
+                    {
+                        alignBB->next = curr;
+                    }
+
+                    alignBB = curr;
+                }
+            }
+#endif
+
```

Comment:
Could you keep a count of aligned loops and decrement the count when you remove a BBF_LOOP_ALIGN bit, then just check for "loopAlignCount > 0" before running the PlaceLoopAlignment phase?

It's really unpleasant to have unrelated phases tied together in a somewhat implicit contract.

Reply:
> Could you keep a count of aligned loops and decrement the count when you remove a BBF_LOOP_ALIGN bit, then just check for "loopAlignCount > 0" before running the PlaceLoopAlignment phase?

Yes, that's exactly what I tried doing, but turns out that sometimes a block/loop is marked as not needing alignment multiple times specially in `AddContainsCallAllContainingLoops()` that miscalculates the count.

Reply:
> in AddContainsCallAllContainingLoops() that miscalculates the count.

Why would it do that? Once you've cleared the bit, you wouldn't decrement again. e.g.,
```
void Compiler::AddContainsCallAllContainingLoops(unsigned lnum)
{

#if FEATURE_LOOP_ALIGN
    // If this is the inner most loop, reset the LOOP_ALIGN flag
    // because a loop having call will not likely to benefit from
    // alignment
    if (optLoopTable[lnum].lpChild == BasicBlock::NOT_IN_LOOP)
    {
        BasicBlock* first = optLoopTable[lnum].lpFirst;
        if (first->isLoopAlign())
        {
            assert(compAlignedLoopCount > 0);
            --compAlignedLoopCount;
            first->bbFlags &= ~BBF_LOOP_ALIGN;
            JITDUMP("Removing LOOP_ALIGN flag for " FMT_LP " that starts at " FMT_BB " because loop has a call.\n", lnum,
                first->bbNum);
         }
    }
#endif
...
```